solr/core/src/java/org/apache/solr/filestore/PackageStoreAPI.java

```diff
@@ -0,0 +1,273 @@
+          if (signatures != null) {
+            vals.put("sig", signatures);
+          }
+          packageStore.put(new PackageStore.FileEntry(buf, new MetaData(vals), path));
+          rsp.add(CommonParams.FILE, path);
+        } catch (IOException e) {
+          throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, e);
+        }
+      } catch (InterruptedException e) {
+        log.error("Unexpected error", e);
+      } catch (KeeperException.NodeExistsException e) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "A write is already in process , try later");
+      } catch (KeeperException e) {
+        log.error("Unexpected error", e);
+      } finally {
+        try {
+          coreContainer.getZkController().getZkClient().delete(TMP_ZK_NODE, -1, true);
+        } catch (Exception e) {
+          log.error("Unexpected error  ", e);
+        }
+      }
+    }
+
+    private List<String> readSignatures(SolrQueryRequest req, ByteBuffer buf)
+        throws SolrException {
+      String[] signatures = req.getParams().getParams("sig");
+      if (signatures == null || signatures.length == 0) return null;
+      List<String> sigs = Arrays.asList(signatures);
+      validate(sigs, buf);
+      return sigs;
+    }
+
+    public void validate(List<String> sigs,
+                         ByteBuffer buf) throws SolrException {
+      Map<String, byte[]> keys = CloudUtil.getTrustedKeys(
+          coreContainer.getZkController().getZkClient(), "exe");
+      if(keys == null || keys.isEmpty()){
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+            "ZK does not have any keys");
+      }
+      CryptoKeys cryptoKeys = null;
+      try {
+        cryptoKeys = new CryptoKeys(keys);
+      } catch (Exception e) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+            "Error parsing public keyts in ZooKeeper");
+      }
+      for (String sig : sigs) {
+       if(cryptoKeys.verify(sig, buf) == null){
+         throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Signature does not match any public key : " +sig );
+       }
+
+      }
+    }
+
+  }
+
+  @EndPoint(
+      path = "/node/files/*",
+      method = SolrRequest.METHOD.GET,
+      permission = PermissionNameProvider.Name.FILESTORE_READ_PERM)
+  public class FSRead {
+    @Command
+    public void read(SolrQueryRequest req, SolrQueryResponse rsp) {
+      String path = req.getPathTemplateValues().get("*");
+      String pathCopy = path;
+      String getFrom = req.getParams().get("getFrom");
+      if (getFrom != null) {
+        coreContainer.getUpdateShardHandler().getUpdateExecutor().submit(() -> {
+          log.debug("Downloading file {}", pathCopy);
+          try {
+            packageStore.fetch(pathCopy, getFrom);
+          } catch (Exception e) {
+            log.error("Failed to download file: " + pathCopy, e);
+          }
+          log.info("downloaded file : {}", pathCopy);
```

Comment:
Lets make this "Downloaded file: {}" (notice capitalization and no space before colon).